move_backwards_recovery/src/move_backwards_recovery.cpp

```diff
@@ -0,0 +1,121 @@
+
+//register this planner as a RecoveryBehavior plugin
+PLUGINLIB_DECLARE_CLASS(move_backwards_recovery, MoveBackRecovery, move_backwards_recovery::MoveBackRecovery, nav_core::RecoveryBehavior)
+
+namespace move_backwards_recovery {
+MoveBackRecovery::MoveBackRecovery(): global_costmap_(NULL), local_costmap_(NULL), 
+  tf_(NULL), initialized_(false), world_model_(NULL){} 
+
+void MoveBackRecovery::initialize(std::string name, tf::TransformListener* tf,
+    costmap_2d::Costmap2DROS* global_costmap, costmap_2d::Costmap2DROS* local_costmap){
+  if(!initialized_){
+    name_ = name;
+    tf_ = tf;
+    global_costmap_ = global_costmap;
+    local_costmap_ = local_costmap;
+    
+    //get parameters from the parameter server
+    ros::NodeHandle private_nh("~/" + name_);
+
+    private_nh.param("distance_backward", distance_backwards_, 0.15);
+    private_nh.param("backwards_velocity", backwards_velocity_, -0.05);
+
+    world_model_ = new base_local_planner::CostmapModel(*local_costmap_->getCostmap());
+    initialized_ = true;
+  }
+  else{
+    ROS_ERROR("You should not call initialize twice on this object, doing nothing");
+  }
+}
+
+MoveBackRecovery::~MoveBackRecovery(){
+  delete world_model_;
+}
+
+void MoveBackRecovery::runBehavior(){
+  if(!initialized_){
+    ROS_ERROR("This object must be initialized before runBehavior is called");
+    return;
+  }
+
+  if(global_costmap_ == NULL || local_costmap_ == NULL){
+    ROS_ERROR("The costmaps passed to the MoveBackRecovery object cannot be NULL. Doing nothing.");
+    return;
+  }
+
+  if(backwards_velocity_ == 0 || distance_backwards_ ==0){
+    ROS_ERROR("Bad input. Either velocity or distance is not setup correctly");
+    return;
+  }
+
+  if(fabs(backwards_velocity_) > 0.1){
+    ROS_ERROR("Velocity is too high to implement move_backwards_recovery");
+    return;
+  }
+
+  // make sure the distance value >0 and backwards velocity < 0
+  distance_backwards_ = (distance_backwards_> 0) ? distance_backwards_ : ((-1) * distance_backwards_) ;
+  backwards_velocity_ = (backwards_velocity_ > 0) ? ((-1)*backwards_velocity_) : backwards_velocity_ ;
+
+  ROS_DEBUG_NAMED("move_backwards_recovery", "MoveBackRecovery started.");
+
+  ros::Rate r(20);
+
+  // setup Nodehandle and publisher
+  ros::NodeHandle n;
+  ros::Publisher vel_pub = n.advertise<geometry_msgs::Twist>(RECOVERY_VELOCITY_CMD, 10);
+
+  tf::Stamped<tf::Pose> global_pose;
+  local_costmap_->getRobotPose(global_pose);
+
+  // the original location
+  double originX = global_pose.getOrigin().x(); 
+  double originY = global_pose.getOrigin().y();   
+
+  double distance_traveled = 0.0;
+
+  while(n.ok()){
+    local_costmap_->getRobotPose(global_pose);
+
+    double cur_x = global_pose.getOrigin().x();
+    double cur_y = global_pose.getOrigin().y();
+    double cur_theta = tf::getYaw(global_pose.getRotation());
+
+    distance_traveled = sqrt((cur_x - originX) * (cur_x - originX) + (cur_y - originY) * (cur_y - originY));
```

Comment:
You may want to just use the squared distance instead of taking the square.  Saves a bit on performance and has the same level of accuracy.

distance_traveled_sq = (cur_x - originX) * (cur_x - originX) + (cur_y - originY) * (cur_y - originY);

float distance_max_sq = distance_backwards_* distance_backwards_;

while (distance_traveled_sq <= distance_max_sq)